### PR TITLE
Budget Splitting

### PR DESCRIPTION
**Pull request by eriktaubeneck**
Earlier this month at the PATCG, we had some conversations about websites/apps being able to delegate to multiple report collectors, and splitting their privacy budget 
across those report collectors. This opens up interesting opportunities:

- Different web platform vendors may allow for different privacy budgets, and websites/apps could allocate the overlapping budget to one report collector, and the 
remaining budget from other vendors to another.
- Different web platform vendors will likely have different sets of helper parties networks that they allow (and potentially even different concepts of private 
computation environments), and this would enable websites/apps to allocate some of their budget to one type of network and some to another.

This PR doesn't address those scenarios specifically (I think they may be better address in the Threat Model document), it does address the ability to allocate budget 
across multiple report collectors.

### CODE REVIEW

**Comment by eriktaubeneck** [comment]
Looking back through some old PRs. @martinthomson and/or @benjaminsavage, could you take a peek at this one?

File: IPA-End-to-End.md [hunk 1]
```diff
@@ -85,6 +86,10 @@ There are a number of different parties that are involved with the protocol in o
 5. _**Attribution Constraint ID**_: An identifier, specified by the _report collector_, to denote if a given pair of _source_ and _trigger events_ can be attributed (beyond having the same _match key_.)
 
 
+### Relationship Between _Websites/Apps_ and _Delegate Report Collectors_
+
+The privacy goals of the IPA proposal are achieved by limiting (through differential privacy) the amount of information about people revealed to individual _websites/apps_. However, many (if not most) _websites/apps_ that use this API will likely do so by working with one or more service providers, which we call _delegated report collectors_. This results in a somewhat complex many-to-many relationship between _websites/apps_ and _delegated report collectors_. For the sake of this document, a _report collector_ is assumed to work with a single _website/app_; in the case where the same service provider works with multiple _websites/apps_, we consider those distinct _report colletors_. Each _report collector_ may be one of many _report collectors_ for that _website/app_, which is primarily meaningful in differential privacy budgeting, [discussed below.](#differential-privacy-budget-management)
```
Comment (martinthomson) [inline IPA-End-to-End.md:91]:
```suggestion
The privacy goals of the IPA proposal are achieved by limiting (through differential privacy) the amount of information about people revealed to individual _websites/apps_. However, many (if not most) _websites/apps_ that use this API will likely do so by working with one or more service providers, which we call _delegated report collectors_. This results in a somewhat complex many-to-many relationship between _websites/apps_ and _delegated report collectors_. For the sake of this document, a _report collector_ is assumed to work with a single _website/app_; in the case where the same service provider works with multiple _websites/apps_, we consider those distinct _report collectors_. Each _report collector_ may be one of many _report collectors_ for that _website/app_, which is primarily meaningful in differential privacy budgeting, [discussed below.](#differential-privacy-budget-management)
```

File: IPA-End-to-End.md [hunk 2]
```diff
@@ -172,15 +177,16 @@ Random noise is added to each breakdown, using ε and the _sensitivity_ to infor
...
 
- In our current approach, we achieve this by providing each _report collector_ with a budget, ε for the given _website/app_ they are querying on behalf of. When _report collectors_ run queries they will specify how much budget to use, ε<sub>i</sub>, which will be deducted from their remaining budget. For example, given an _epoch_ limit of ε, a _report collector_ could perform 10 queries, each with global DP applied at ε/10, or a more complicated set of queries such as three with ε/5 and four with ε/10. (Note that smaller ε<sub>i</sub> result in _more_ noise and, thus, _more_ privacy.)
+When _report collectors_ run queries they will specify how much budget to use, ε<sub>i</sub>, which will be deducted from their remaining budget. For example, given an _epoch_ limit of ε, a _report collector_ could perform 10 queries, each with global DP applied at ε/10, or a more complicated set of queries such as three with ε/5 and four with ε/10. (Note that smaller ε<sub>i</sub> result in _more_ noise and, thus, _more_ privacy.)
```
Comment (martinthomson) [inline IPA-End-to-End.md:184]: Does this paragraph need to use $\epsilon_r$ ?

File: IPA-End-to-End.md [hunk 3]
```diff
@@ -172,15 +177,16 @@ Random noise is added to each breakdown, using ε and the _sensitivity_ to infor
...
 
-The _helper party network_ will need to maintain this budget, per _website/app_, over the course of an _epoch_, preventing _report collectors_ from issuing additional queries once that budget is exhausted. At the beginning of the next _epoch_, every _report collector’s_ budget will refresh to ε. Note that in infinite time, the information revealed also goes to infinity; our aim here is to control how quickly that divergence happens.
+The _helper party network_ will need to maintain this budget, per _report collector_ and reflecting the commitments made by _websites/apps_, over the course of an _epoch_, preventing _report collectors_ from issuing additional queries once that budget is exhausted. At the beginning of the next _epoch_, every _report collector’s_ budget will refresh to ε<sub>r</sub> (baring any changes in allocation by the respective _website/app_.) Note that in infinite time, the information revealed also goes to infinity; our aim here is to control how quickly that divergence happens.
```
Comment (martinthomson) [inline IPA-End-to-End.md:188]:
```suggestion
The _helper party network_ will need to maintain this budget, per _report collector_ and reflecting the commitments made by _websites/apps_, over the course of an _epoch_, preventing _report collectors_ from issuing additional queries once that budget is exhausted. At the beginning of the next _epoch_, every _report collector’s_ budget will refresh to ε<sub>r</sub> (barring any changes in allocation by the respective _website/app_.) Note that in infinite time, the information revealed also goes to infinity; our aim here is to control how quickly that divergence happens.
```

**Review by martinthomson — APPROVED** [review]
typos only

File: IPA-End-to-End.md (hunk 2 above)
Comment (martinthomson) [inline IPA-End-to-End.md:184]:
```suggestion
When _report collectors_ run queries they will specify how much budget to use, ε<sub>i</sub>, which will be deducted from their remaining budget. For example, given an _epoch_ limit of ε, a _report collector_ could perform 10 queries, each with global DP applied at ε/10, or a more complicated set of queries such as three with ε/5 and four with ε/10. When split over multiple queries, smaller ε<sub>i</sub> values result in _more_ noise on each query; each query has greater privacy protection, but the overall privacy loss is similar.
```

Comment (eriktaubeneck) [reply IPA-End-to-End.md]: it's probably clearer, yes. I'll update accordingly

File: IPA-End-to-End.md [hunk 4]
```diff
@@ -172,15 +177,16 @@ Random noise is added to each breakdown, using ε and the _sensitivity_ to infor
...
 
- In our current approach, we achieve this by providing each _report collector_ with a budget, ε for the given _website/app_ they are querying on behalf of. When _report collectors_ run queries they will specify how much budget to use, ε<sub>i</sub>, which will be deducted from their remaining budget. For example, given an _epoch_ limit of ε, a _report collector_ could perform 10 queries, each with global DP applied at ε/10, or a more complicated set of queries such as three with ε/5 and four with ε/10. (Note that smaller ε<sub>i</sub> result in _more_ noise and, thus, _more_ privacy.)
+When _report collectors_ run queries they will specify how much budget to use, ε<sub>i</sub>, which will be deducted from their remaining budget. For example, given an _epoch_ limit of ε, a _report collector_ could perform 10 queries, each with global DP applied at ε/10, or a more complicated set of queries such as three with ε/5 and four with ε/10. When split over multiple queries, smaller ε<sub>i</sub> values result in _more_ noise on each query; each query has greater privacy protection, but the overall privacy loss is similar.
```
Comment (eriktaubeneck) [inline IPA-End-to-End.md:184]:
```suggestion
When _report collectors_ run queries they will specify how much budget to use, ε<sub>i</sub>, which will be deducted from their remaining budget. For example, given an _epoch_ limit of ε<sub>r</sub>, a _report collector_ could perform 10 queries, each with global DP applied at ε<sub>r</sub>/10, or a more complicated set of queries such as three with ε<sub>r</sub>/5 and four with ε<sub>r</sub>/10. When split over multiple queries, smaller ε<sub>i</sub> values result in _more_ noise on each query; each query has greater privacy protection, but the overall privacy loss is similar.
```

**Pull request merged**